src/anonymous-credentials/schema.ts

```diff
@@ -587,56 +624,78 @@ export class CredentialSchema extends Versioned {
-      if (typeof node.properties !== 'undefined') {
-        const result: object = {};
-        const keys = Object.keys(node.properties);
-        keys.forEach((k) => {
-          const value = node.properties[k];
-
-          let insVal = {};
-          if (value.type === 'object') {
-            if (typeof value.properties === 'object') {
-              insVal = CredentialSchema.flattenJSONSchema(value);
-            } else {
-              throw new Error(`${k} must have properties field that is an object`);
-            }
-          } else if (value.type === 'array') {
-            if (Array.isArray(value.items)) {
-              insVal = value.items.map((i) => CredentialSchema.flattenJSONSchema(i));
-            } else {
-              throw new Error('No indefinite length array support');
-            }
+    // Will either have a "type" property or will be defined using "$ref"
+    const typ = node.type;
+    const ref = node.$ref;
+    if (typ !== undefined) {
+      switch (typ) {
+        case 'string':
+          return node;
+        case 'integer':
+          // If key "minimum" is not present or not an integer, validation logic in other function will throw.
+          return { type: this.INT_TYPE, minimum: node.minimum };
+        case 'number':
+          return { type: this.NUM_TYPE, minimum: node.minimum, decimalPlaces: getDecimalPlaces(node.multipleOf) };
+        case 'object':
+          if (node.properties !== undefined) {
+            const result = {};
+            Object.entries(node.properties).forEach(([k, v]) => {
+              result[k] = CredentialSchema.convertToInternalSchemaObj(v, createFullName(nodeKeyName, k));
+            });
+            return result;
           } else {
-            insVal = value;
+            throw new Error(`Schema object key ${nodeKeyName} must have properties object`);
           }
-
-          result[k] = insVal;
-        });
-        return result;
-      } else {
-        throw new Error('Schema object must have properties object');
+        case 'array':
+          return node.items.map((i) => CredentialSchema.convertToInternalSchemaObj(i, createFullName(nodeKeyName, i)));
+        default:
+          throw new Error(`Unknown type for key ${nodeKeyName} in schema: ${typ}`);
+      }
+    } else if (ref !== undefined) {
+      switch (ref) {
+        case '#/definitions/encryptableString':
+          return { type: this.STR_REV_TYPE, compress: false };
+        case '#/definitions/encryptableCompString':
+          return { type: this.STR_REV_TYPE, compress: true };
+        case '#/definitions/positiveInteger':
+          return { type: this.POSITIVE_INT_TYPE };
+        default:
+          throw new Error(`Unknown ref for key ${nodeKeyName} in schema: ${ref}`);
       }
-    } else if (node.type === 'array') {
-      return node.items.map((i) => CredentialSchema.flattenJSONSchema(i));
     } else {
-      return node;
+      if (Array.isArray(node.allOf)) {
```

Comment:
what is the intention here? allOf requires that a value conforms to all schemas within it - but that's not what this code and lines below are doing

Reply:
The intention is to parse `allOf` of this form

```
"allOf": [
                { "$ref": "#/definitions/positiveNumber" },
                { "multipleOf": 0.1 }
            ]
```

If these are not found or not in the format expected, an error will be thrown while creating `CredentialSchema` obj. Any extra properties don't impact the parsing logic.

This is the exact format i had shared in slack before stating this work.

Reply:
we discussed on the call about how allof is potentially confusing given its behaviour here is not what would be expected in jsonschema, and may not work for every possible case + since we arent relying on refs to define positive numbers (they still can if you prefer that for your schemas) as detecitng minimum value was chosen